apps/server/lib/http/auth.ts

```diff
@@ -0,0 +1,75 @@
+import {
+	JellyfishKernel,
+	SessionContract,
+} from '@balena/jellyfish-types/build/core';
+import bcrypt from 'bcrypt';
+import _ from 'lodash';
+
+// TODO: Move this definition elsewhere, and inject it into the middleware.
+const sessionIdAndTokenValidator =
+	(context, jellyfish: JellyfishKernel) =>
+	async (session: string, sessionToken?: string): Promise<boolean> => {
+		try {
+			const card = await jellyfish.getCardById<SessionContract>(
+				context,
+				jellyfish.sessions!.admin,
+				session,
+			);
+			return bcrypt.compare(sessionToken, card!.data.token!.authentication);
+		} catch (e) {
+			return false;
+		}
+	};
+
+// TODO: Move this definition elsewhere, and inject it into the middleware.
+const sessionIdValidator =
+	(context, jellyfish: JellyfishKernel) =>
+	async (session: string): Promise<boolean> => {
+		try {
+			const card = jellyfish.getCardById<SessionContract>(
+				context,
+				jellyfish.sessions!.admin,
+				session,
+			);
+			if (!card) {
+				return false;
+			}
+			return true;
+		} catch (e) {
+			return false;
+		}
+	};
+
+export const authMiddleware =
+	(jellyfish: JellyfishKernel, options: { guestSession: string }) =>
+	async (request, _response, next) => {
+		const sessionHeader = request.headers['x-session-id'];
+		const authorizationHeader = request.headers['authorization'];
+
+		const authorizationHeaderValue = _.last(_.split(authorizationHeader, ' '));
+
+		request.session = options.guestSession;
+
+		// TODO: take validation function injected via options rather than building them here.
+		const validateSessionId = sessionIdValidator(request.context, jellyfish);
+		const validateSessionIdAndToken = sessionIdAndTokenValidator(
+			request.context,
+			jellyfish,
+		);
+
+		if (sessionHeader) {
```

Comment:
We should probably check for `sessionHeader` etc earlier on so we don't unnecessarily query the DB with the validation functions.